File: android/feature/photos/src/main/kotlin/xyz/klinik/feature/photos/PhotoGalleryModel.kt
package xyz.klinik.feature.photos

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import xyz.klinik.network.ApiError
import xyz.klinik.network.ClinicalPhoto
import xyz.klinik.network.GalleryGroup
import xyz.klinik.network.PhotoCategory
import xyz.klinik.network.PhotosApi
import xyz.klinik.network.UiText
import xyz.klinik.network.messageKey
import xyz.klinik.network.uiText

sealed interface GalleryPhase {
    data object Loading : GalleryPhase
    data object Loaded : GalleryPhase

    /** No photos yet. Not a failure. */
    data object Empty : GalleryPhase

    data object NotFound : GalleryPhase
    data class Failed(val messageKey: String) : GalleryPhase
}

/** The two photos a comparison shows, and which way round they go. */
data class ComparisonPair(val before: ClinicalPhoto, val after: ClinicalPhoto)

data class GalleryState(
    val phase: GalleryPhase = GalleryPhase.Loading,
    val groups: List<GalleryGroup> = emptyList(),
    val selectedArea: String? = null,
    val uploading: Boolean = false,
    val error: UiText? = null,
) {
    val selectedGroup: GalleryGroup?
        get() = groups.firstOrNull { it.id == selectedArea } ?: groups.firstOrNull()

    /**
     * The pair a comparison opens on: the earliest and the latest of the
     * selected body area.
     *
     * Null for a single photo, because a slider over one image does nothing and
     * offering it suggests there is a change to see.
     */
    val comparison: ComparisonPair?
        get() {
            val photos = selectedGroup?.photos ?: return null
            if (photos.size < 2) return null

            return ComparisonPair(photos.first(), photos.last())
        }
}

/** The before/after gallery (spec M7). */
class PhotoGalleryModel(
    private val api: PhotosApi,
    private val patientId: String,
) {
    private val _state = MutableStateFlow(GalleryState())
    val state: StateFlow<GalleryState> = _state.asStateFlow()

    private val uploadLock = Mutex()

    suspend fun load(category: PhotoCategory? = null) {
        _state.value = _state.value.copy(phase = GalleryPhase.Loading)

        try {
            val groups = api.gallery(patientId, category)

            _state.value = _state.value.copy(
                groups = groups,
                selectedArea = groups.firstOrNull()?.id,
                phase = if (groups.isEmpty()) GalleryPhase.Empty else GalleryPhase.Loaded,
            )
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                phase = if (error is ApiError.NotFound) {
                    GalleryPhase.NotFound
                } else {
                    GalleryPhase.Failed((error as? ApiError)?.messageKey() ?: "error.server")
                },
            )
        }
    }

    fun select(area: String) {
        _state.value = _state.value.copy(selectedArea = area)
    }

    /** The photo a new capture lines up against, for the translucent guide. */
    suspend fun overlayReference(bodyArea: String): ClinicalPhoto? =
        runCatching { api.overlayReference(patientId, bodyArea) }.getOrNull()

    suspend fun upload(
        path: String,
        filename: String,
        category: PhotoCategory,
        bodyArea: String? = null,
        phaseLabel: String? = null,
    ): Boolean {
        if (!uploadLock.tryLock()) return false

        _state.value = _state.value.copy(uploading = true, error = null)

        try {
            api.upload(patientId, path, filename, category, bodyArea, phaseLabel)
        } catch (error: Throwable) {
            // The server refuses a format whose location data it cannot strip,
            // and says so. Replacing that with our own message would leave the
            // person guessing why a perfectly good photo was rejected.
            _state.value = _state.value.copy(
                uploading = false,
                error = (error as? ApiError)?.uiText() ?: UiText.Key("error.server"),
            )
            return false
        } finally {
            uploadLock.unlock()
        }

        // Reloaded rather than appended: the server decides the stored type and
        // the grouping, and a guessed row would disagree after the next refresh.
        load()
        _state.value = _state.value.copy(uploading = false)
        return true
    }

    suspend fun remove(photoId: String) {
        try {
            api.remove(photoId)
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                error = (error as? ApiError)?.uiText() ?: UiText.Key("error.server"),
            )
            return
        }

        load()
    }
}

File: ios/Sources/KlinikPhotosFeature/PhotoGalleryModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum GalleryPhase: Sendable, Equatable {
    case loading
    case loaded
    /// No photos yet. Not a failure.
    case empty
    case notFound
    case failed(String)
}

/// The two photos a comparison shows, and which way round they go.
public struct ComparisonPair: Sendable, Equatable {
    public let before: ClinicalPhoto
    public let after: ClinicalPhoto
}

public struct GalleryState: Sendable, Equatable {
    public var phase: GalleryPhase = .loading
    public var groups: [GalleryGroup] = []
    public var selectedArea: String?
    public var uploading = false
    public var error: String?

    /**
     * The last photograph could not be sent and is being held (spec M15).
     *
     * Not an error: the bytes are on the phone and will go when there is a
     * connection. A wound photograph taken in a hotel with no signal used to
     * be lost here with a red message.
     */
    public var queued = false

    public var selectedGroup: GalleryGroup? {
        groups.first { $0.id == selectedArea } ?? groups.first
    }

    /**
     * The pair a comparison opens on: the earliest and the latest of the
     * selected body area.
     *
     * Nil for a single photo, because a comparison of one image against itself
     * is a slider that does nothing, and showing it suggests there is a change
     * to see.
     */
    public var comparison: ComparisonPair? {
        guard let photos = selectedGroup?.photos, photos.count >= 2,
              let first = photos.first, let last = photos.last
        else { return nil }

        return ComparisonPair(before: first, after: last)
    }

    public init() {}
}

/// The before/after gallery (spec M7).
public actor PhotoGalleryModel {
    private let api: PhotosAPI
    private let subject: RecordSubject
    /// Where a photograph goes when the connection cannot carry it. Nil in
    /// tests that have nothing to say about it.
    private let queue: PendingUploadQueue?

    private(set) public var state = GalleryState()

    public init(api: PhotosAPI, subject: RecordSubject, queue: PendingUploadQueue? = nil) {
        self.api = api
        self.subject = subject
        self.queue = queue
    }

    /**
     * Hands a photograph the connection could not carry to the queue.
     *
     * Reports success, because from the patient's side the photograph *is*
     * taken — what is outstanding is the delivery, and the screen says so.
     * With no queue attached the failure is reported exactly as before.
     */
    private func hold(
        fileURL: URL,
        category: PhotoCategory,
        bodyArea: String?,
        phaseLabel: String?,
        failure: APIError
    ) async -> Bool {
        guard let queue else {
            state.error = L10n.message(for: failure)
            return false
        }

        var fields = ["category": category.rawValue]
        if let bodyArea { fields["bodyArea"] = bodyArea }
        if let phaseLabel { fields["phaseLabel"] = phaseLabel }

        do {
            try await queue.keepPhoto(
                fileURL: fileURL,
                subject: subject,
                contentType: "image/jpeg",
                fields: fields
            )
        } catch {
            // The queue is the only reason to claim the photograph is safe.
            state.error = L10n.message(for: failure)
            return false
        }

        state.queued = true

        return true
    }

    public func currentState() -> GalleryState { state }

    public func load(category: PhotoCategory? = nil) async {
        state.phase = .loading

        do {
            let groups = try await api.gallery(subject: subject, category: category)
            state.groups = groups
            state.selectedArea = groups.first?.id
            state.phase = groups.isEmpty ? .empty : .loaded
        } catch let error as APIError {
            if case .notFound = error {
                state.phase = .notFound
            } else {
                state.phase = .failed(L10n.message(for: error))
            }
        } catch {
            state.phase = .failed(L10n.string("error.server"))
        }
    }

    public func select(area: String) {
        state.selectedArea = area
    }

    /// The photo a new capture lines up against, for the translucent guide.
    public func overlayReference(bodyArea: String) async -> ClinicalPhoto? {
        try? await api.overlayReference(subject: subject, bodyArea: bodyArea)
    }

    @discardableResult
    public func upload(
        fileURL: URL,
        category: PhotoCategory,
        bodyArea: String?,
        phaseLabel: String?
    ) async -> Bool {
        guard !state.uploading else { return false }

        state.uploading = true
        state.error = nil
        state.queued = false
        defer { state.uploading = false }

        do {
            _ = try await api.upload(
                subject: subject,
                fileURL: fileURL,
                category: category,
                bodyArea: bodyArea,
                phaseLabel: phaseLabel
            )
        } catch let error as APIError where error.isConnectivity && queue != nil {
            return await hold(
                fileURL: fileURL,
                category: category,
                bodyArea: bodyArea,
                phaseLabel: phaseLabel,
                failure: error
            )
        } catch let error as APIError {
            // The server refuses a format whose location data it cannot strip,
            // and says so. Replacing that with our own message would leave the
            // person guessing why a perfectly good photo was rejected.
            state.error = L10n.message(for: error)
            return false
        } catch {
            state.error = L10n.string("error.server")
            return false
        }

        // Reloaded rather than appended: the server decides the stored type and
        // the phase grouping, and a guessed row would disagree with the one
        // beside it after the next refresh.
        await load()
        return true
    }

    public func remove(photoId: String) async {
        do {
            try await api.remove(photoId: photoId)
        } catch let error as APIError {
            state.error = L10n.message(for: error)
            return
        } catch {
            state.error = L10n.string("error.server")
            return
        }

        await load()
    }
}
